File: leetcode/java/0023. Merge k Sorted Lists.java
package leetcode.java;

import java.util.LinkedList;
import java.util.Queue;

// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

// Time complexity: O(kN), where k is the number of linked lists and N is the total number of nodes in lists
// Space complexity: O(N)

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) {
            return null;
        }
        Queue<ListNode> q = new LinkedList<>();
        for (ListNode l : lists) {
            q.add(l);
        }

        while (q.size() > 1) {
            ListNode l1 = q.poll();
            ListNode l2 = q.poll();
            q.add(this.merge(l1, l2));
        }

        return q.poll();
    }

    public ListNode merge(ListNode l1, ListNode l2) {
        ListNode res = new ListNode();
        ListNode temp = res;
        while (l1 != null && l2 != null) {
            if (l1.val > l2.val) {
                temp.next = new ListNode(l2.val);
                l2 = l2.next;
            } else {
                temp.next = new ListNode(l1.val);
                l1 = l1.next;
            }
            temp = temp.next;
        }
        while (l1 != null) {
            temp.next = new ListNode(l1.val);
            l1 = l1.next;
            temp = temp.next;
        }
        while (l2 != null) {
            temp.next = new ListNode(l2.val);
            l2 = l2.next;
            temp = temp.next;
        }
        return res.next;
    }
}

File: leetcode/python/0023. Merge k Sorted Lists.py
import heapq

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Time complexity: O(NlogK), where K is the number of linked lists and N is the total number of nodes in lists
# Space complexity: O(N)

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """ 
        heap = []
        res = ListNode()
        temp = res
        for l in lists:
            if l is not None:
                heap.append([l.val, l.next])
        heapq.heapify(heap)
        while (len(heap) > 0):
            val, node = heapq.heappop(heap)
            if node is not None:
                heapq.heappush(heap, [node.val, node.next])
            temp.next = ListNode(val)
            temp = temp.next
        return res.next